### PR TITLE
Plugins: fetch variant features from VEP cache or Ensembl API

### PR DESCRIPTION
**Pull request by nuno-agostinho**
ENSVAR-6290: new subroutines to allow plugins to fetch variant features using:
- (indexed or non-indexed) VEP cache if `--cache` is enabled.
- Otherwise, Ensembl API if `--database` is enabled.

You can test the changes here using PR https://github.com/Ensembl/VEP_plugins/pull/723. Please test using:
* Indexed VEP cache.
* Non-indexed VEP cache (slower).
* Database mode.

### CODE REVIEW

File: modules/Bio/EnsEMBL/Variation/Utils/BaseVepPlugin.pm [hunk 1]
```diff
@@ -326,6 +326,160 @@ sub check_variant_feature_type {
...
+  Returntype : arrayref of Bio::EnsEMBL::Variation::VariationFeature
+  Status     : Experimental
+
+=cut
+
+sub _fetch_cache_vars {
+  my ($self, $chr, $start, $end) = @_;
+
+  my ($as, $variants);
+  if (defined($as = $self->_get_AnnotationSource('VariationTabix'))) {
```
Comment (nakib103) [inline modules/Bio/EnsEMBL/Variation/Utils/BaseVepPlugin.pm:378]: Hi @nuno-agostinho , should this `_get_AnnotationSource` should also be part of `BaseVepPlugin.pm`. It seems to generic to be in Paralogues plugin only

File: modules/Bio/EnsEMBL/Variation/Utils/BaseVepPlugin.pm [hunk 2]
```diff
@@ -326,6 +326,160 @@ sub check_variant_feature_type {
...
+
+    while(my $line = $iter->next) {
+      chomp $line;
+      my $var = $as->parse_variation($line);
+      push @$variants, $self->_create_vf($var);
+    }
+  } elsif (defined($as = $self->_get_AnnotationSource('Variation'))) {
+    warn("Using non-indexed VEP cache is slow; for optimal performance, please use indexed VEP cache\n")
+      unless $self->{slow_warning};
+    $self->{slow_warning} = 1;
```
Comment (nakib103) [inline modules/Bio/EnsEMBL/Variation/Utils/BaseVepPlugin.pm:395]:
```suggestion
    warn("Using non-indexed VEP cache is slow; for optimal performance, please use indexed VEP cache\n")
      unless $self->{skip_slow_warning};
    $self->{skip_slow_warning} = 1;
```

**Review by nakib103 — CHANGES_REQUESTED** [review]

File: modules/Bio/EnsEMBL/Variation/Utils/BaseVepPlugin.pm [hunk 3]
```diff
@@ -326,6 +326,160 @@ sub check_variant_feature_type {
...
+    for my $region (@regions) {
+      my ($c, $s) = @$region;
+
+      my $file = $as->get_dump_file_name(
+        $c,
+        ($s * $cache_region_size) + 1,
+        ($s + 1) * $cache_region_size
+      );
+      next unless -e $file;
+      my $gz = gzopen($file, 'rb');
```
Comment (nakib103) [inline modules/Bio/EnsEMBL/Variation/Utils/BaseVepPlugin.pm:411]: Need to use `use Compress::Zlib;` for `gzopen` function to work.

File: modules/Bio/EnsEMBL/Variation/Utils/BaseVepPlugin.pm [hunk 4]
```diff
@@ -326,6 +326,160 @@ sub check_variant_feature_type {
...
+
+      my $line;
+      while($gz->gzreadline($line)) {
+        chomp $line;
+        # ignore non-overlapping variants
+        my $var = $as->parse_variation($line);
+        next unless overlap($start, $end, $var->{start}, $var->{end});
```
Comment (nakib103) [inline modules/Bio/EnsEMBL/Variation/Utils/BaseVepPlugin.pm:418]: @nuno-agostinho, `overlap` function is not defined here or imported, hence not working.

**Review by nakib103 — CHANGES_REQUESTED** [review]

File: modules/Bio/EnsEMBL/Variation/Utils/BaseVepPlugin.pm [hunk 5]
```diff
@@ -326,6 +326,160 @@ sub check_variant_feature_type {
...
+        ($s + 1) * $cache_region_size
+      );
+      next unless -e $file;
+      my $gz = gzopen($file, 'rb');
+
+      my $line;
+      while($gz->gzreadline($line)) {
+        chomp $line;
+        # ignore non-overlapping variants
+        my $var = $as->parse_variation($line);
```
Comment (nakib103) [inline modules/Bio/EnsEMBL/Variation/Utils/BaseVepPlugin.pm:417]: the non-indexed cache lines do not contain the chromosome, we should add it manually.

```suggestion
        my $var = $as->parse_variation($line);
        $var->{chr} = $c;
```

**Review by nakib103 — CHANGES_REQUESTED** [review]

**Comment by nuno-agostinho** [comment]
The current design of ensembl-vep and ensembl-variation make @nakib103's suggestions not trivial to implement.

As such, I will close this PR. All changes in this PR were now included in https://github.com/Ensembl/VEP_plugins/pull/723.

The changes proposed here will be reconsidered for the next version of VEP.

**Pull request closed without merging**